[Intel GPU] Add XPU memory-related APIs

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129919
* #130923

# Motivation
According to https://github.com/pytorch/pytorch/issues/116322, we will help unify the device allocator. So we introduce a simple xpu device allocator only with the key functionality first. And expect to add some memory statistics-related functionality after the unification. 
But now, some memory statistic-related APIs listed in https://github.com/pytorch/pytorch/issues/127929 are requested. We need more time to unify the device allocator. In order to facilitate the user experience, we expect to support these memory statistic-related APIs before the unification.

# Additional Context
Fixes: #127929

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

guangyey committed:
Add XPU memory-related APIs

[ghstack-poisoned]

guangyey committed:
Update on "Add XPU memory-related APIs"


# Motivation
According to https://github.com/pytorch/pytorch/issues/116322, we will help unify the device allocator. So we introduce a simple xpu device allocator only with the key functionality first. And expect to add some memory statistics-related functionality after the unification. 
But now, some memory statistic-related APIs listed in https://github.com/pytorch/pytorch/issues/127929 are requested. We need more time to unify the device allocator. In order to facilitate the user experience, we expect to support these memory statistic-related APIs before the unification.


[ghstack-poisoned]

dvrogozh commented on c10/xpu/XPUCachingAllocator.cpp:
```diff
@@ -26,6 +26,49 @@ constexpr size_t kRoundLarge = 2097152;
+  stat.current -= static_cast<int64_t>(amount);
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+      stat.current >= 0,
+      "Negative tracked stat in CUDA allocator (likely logic error).");
```
```suggestion
      "Negative tracked stat in XPU allocator (likely logic error).");
```

guangyey replied:
updated. Thanks.

dvrogozh commented:
Add to PR description to automatically close the issue on merge:
```
Fixes: #127929
```

dvrogozh reviewed (approved):
looks good

guangyey committed:
Update on "Add XPU memory-related APIs"


# Motivation
According to https://github.com/pytorch/pytorch/issues/116322, we will help unify the device allocator. So we introduce a simple xpu device allocator only with the key functionality first. And expect to add some memory statistics-related functionality after the unification. 
But now, some memory statistic-related APIs listed in https://github.com/pytorch/pytorch/issues/127929 are requested. We need more time to unify the device allocator. In order to facilitate the user experience, we expect to support these memory statistic-related APIs before the unification.


[ghstack-poisoned]

EikanWang commented:
@pytorchbot rebase

guangyey commented:
@pytorchbot rebase -b main

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

EikanWang commented on c10/xpu/XPUCachingAllocator.cpp:
Most of the changes are the same as the CUDA Caching Allocator. Should we abstract these utility functions and reuse these functions for both CUDA and XPU?

guangyey replied:
I will unify the device allocator for `cuda` and `xpu` in a separate PR.

EikanWang commented on c10/xpu/XPUCachingAllocator.h:
```diff
@@ -5,12 +5,46 @@
 
 namespace c10::xpu::XPUCachingAllocator {
 
+struct Stat {
+  int64_t current = 0;
+  int64_t peak = 0;
+  int64_t allocated = 0;
+  int64_t freed = 0;
+};
+
+enum struct StatType : uint64_t {
+  AGGREGATE = 0,
+  SMALL_POOL = 1,
+  LARGE_POOL = 2,
+  NUM_TYPES = 3 // remember to update this whenever a new stat type is added
+};
+
+typedef std::array<Stat, static_cast<size_t>(StatType::NUM_TYPES)> StatArray;
+
+// Struct containing memory allocator summary statistics for a device.
+struct DeviceStats {
+  // SUM: bytes allocated by this memory allocator
+  StatArray allocated_bytes;
+  // SUM: bytes reserved by this memory allocator (both free and used)
+  StatArray reserved_bytes;
+  // SUM: bytes within active memory blocks
+  StatArray active_bytes;
+  // SUM: bytes requested by client code
+  StatArray requested_bytes;
+};
+
```
ditto

EikanWang commented:
@pytorchbot rebase

guangyey committed:
Update

[ghstack-poisoned]

dvrogozh commented:
What's the ETA to get this merged? These APIs block running some HF tests which might help to uncover other important issues.

abhilash1910 reviewed (approved):
LGTM! This was indeed needed. Thanks @guangyey

dvrogozh commented:
@guangyey : can you, please, update status of this PR? When it will be merged? If not, what's pending? As of now it's not clear from the comments above. You plan some refactoring done after this PR merged, but it seems no actions taken to actually merge it. Maintainers are not pinged for review.

guangyey commented:
Hi @dvrogozh , this PR is ready to review. I think Alban will review it when it is in his queue of regular PRs.

guangyey committed:
Update

[ghstack-poisoned]

dvrogozh commented:
@albanD : can you, please, help review this PR? that's long standing PR with key uAPIs heavily used across ecosystem projects.

guangyey committed:
Update

[ghstack-poisoned]

EikanWang commented:
@guangyey , could you help rebase the code and pass the CI?

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

dvrogozh commented:
@albanD : can this PR, please, be reviewed? It would be nice to have it in upcoming pytorch 2.5 release since it blocks running some workloads and tests with huggingface.

albanD commented:
This is pretty hard to review since I am not familiar with this code. I guess this is copy pasted from the cuda side?
What is the plan to align these APIs with other devices?

guangyey commented:
> This is pretty hard to review since I am not familiar with this code. I guess this is copy pasted from the cuda side?

Yes, this code logic is the same as cuda.

guangyey commented:
> What is the plan to align these python APIs with other devices?

In long term, we intend to give a generic python memory-related API design [[RFC] A device-agnostic Python device memory related API design for stream-based accelerators](https://github.com/pytorch/pytorch/issues/134978) on top of https://github.com/pytorch/pytorch/issues/116322.
On summary is 
- abstract C++ device caching allocator
- unify the C++ API via the abstract, like `GetAllocator(device_type)->emptyCache()`
- unify the python frontend API via C++ API, like `torch.empty_cache`.

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

EikanWang commented:
> This is pretty hard to review since I am not familiar with this code. I guess this is copy pasted from the cuda side? What is the plan to align these python APIs with other devices?

Hi @albanD , As @guangyey mentioned, this logic and implementation is similar with CUDA. And this PR intends to serve HuggingFace Intel GPU support because some HuggingFace models leverage these APIs to get device memory information. 

Regarding python APIs alignment, do you think it makes sense to define a staging plan because I think it may take more time conclude the final solution and implementation?

- Stage 0: The runtime-related APIs to be implemented remain device-specific. Moreover, these APIs should meet actual requirements, rather than merely aiming for comprehensive coverage.
- Stage 1: Align runtime related python APIs with other devices

Are you okay with it? @albanD

albanD commented on torch/xpu/__init__.py:
```diff
@@ -448,6 +435,7 @@ def _get_rng_state_offset(device: Union[int, str, torch.device] = "xpu") -> int:
     return default_generator.get_offset()
 
 
+from .memory import *  # noqa: F403
```
nit: explicit imports are better for typing and readability. Would be nice to fix that going forward and remove the noqa

guangyey replied:
Thanks very much. I removed `noqa` in this PR.

albanD reviewed (approved):
Feel free to cleanup the import in a follow up PR.

Sounds ok as a temporary solution (didn't carefully review the logic). I do expect that if the goal is to support HF and other third party, the unified API is even more urgent as otherwise it's going to be a never ending battle to update every single one of these libraries.

EikanWang commented:
> I do expect that if the goal is to support HF and other third party, the unified API is even more urgent as otherwise it's going to be a never ending battle to update every single one of these libraries.

Thanks, @albanD . And I could not agree more.

guangyey commented:
> Feel free to cleanup the import in a follow up PR.
> 
> Sounds ok as a temporary solution (didn't carefully review the logic). I do expect that if the goal is to support HF and other third party, the unified API is even more urgent as otherwise it's going to be a never ending battle to update every single one of these libraries.

Yes, I agree. I will prioritize and focus on these unified API tasks.

guangyey committed:
Update

[ghstack-poisoned]

guangyey committed:
Update

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge
